lib/test/clojure_lsp/dep_graph_test.clj

```diff
@@ -0,0 +1,259 @@
+      (is (not (:from-internal? xxx-yyy)))))
+  (testing "adding dependency"
+    (h/clean-db!)
+    (load-code "/bbb.clj"
+               "(ns bbb)")
+    (load-code "/aaa.clj"
+               "(ns aaa)")
+    (let [db @db/db*]
+      (is (empty? (get-in db [:dep-graph 'aaa :dependencies])))
+      (is (empty? (get-in db [:dep-graph 'bbb :dependents]))))
+    (load-code "/aaa.clj"
+               "(ns aaa"
+               " (:require [bbb :as b]))")
+    (let [db @db/db*]
+      (is (get-in db [:dep-graph 'aaa :dependencies 'bbb]))
+      (is (get-in db [:dep-graph 'bbb :dependents 'aaa]))))
+  (testing "removing dependency"
+    (h/clean-db!)
+    (load-code "/bbb.clj"
+               "(ns bbb)")
+    (load-code "/aaa.clj"
+               "(ns aaa"
+               " (:require [bbb :as b]))")
+    (let [db @db/db*]
+      (is (get-in db [:dep-graph 'aaa :dependencies 'bbb]))
+      (is (get-in db [:dep-graph 'bbb :dependents 'aaa])))
+    (load-code "/aaa.clj"
+               "(ns aaa)")
+    (let [db @db/db*]
+      (is (empty? (get-in db [:dep-graph 'aaa :dependencies])))
+      (is (empty? (get-in db [:dep-graph 'bbb :dependents])))))
+  (testing "removing duplicate dependency"
+    (h/clean-db!)
+    (load-code "/bbb.clj"
+               "(ns bbb)")
+    (load-code "/aaa.clj"
+               "(ns aaa"
+               " (:require [bbb :as b]"
+               "           [bbb :as b2]))")
+    (let [db @db/db*]
+      (is (get-in db [:dep-graph 'aaa :dependencies 'bbb]))
+      (is (get-in db [:dep-graph 'bbb :dependents 'aaa]))
+      (is (= '{b 1, b2 1} (get-in db [:dep-graph 'bbb :aliases]))))
+    (load-code "/aaa.clj"
+               "(ns aaa"
+               " (:require [bbb :as b]))")
+    (let [db @db/db*]
+      (is (get-in db [:dep-graph 'aaa :dependencies 'bbb]))
+      (is (get-in db [:dep-graph 'bbb :dependents 'aaa]))
+      (is (= '{b 1} (get-in db [:dep-graph 'bbb :aliases])))))
+  (testing "deleting file"
+    (h/clean-db!)
+    (load-code "/bbb.clj"
+               "(ns bbb)")
+    (load-code "/aaa.clj"
+               "(ns aaa"
+               " (:require [bbb :as b]))")
+    (load-code "/ccc.clj"
+               "(ns ccc"
+               " (:require [aaa :as a]))")
+    (let [db @db/db*]
+      (is (seq (get-in db [:dep-graph 'aaa :dependencies])))
+      (is (= '{ccc 1} (get-in db [:dep-graph 'aaa :dependents])))
+      (is (seq (get-in db [:dep-graph 'aaa :filenames])))
+      (is (seq (get-in db [:dep-graph 'bbb :dependents])))
+      (is (not (nil? (get-in db [:file-meta "/aaa.clj"])))))
+    #_(alter-var-root #'db/diagnostics-chan (constantly (async/chan 1)))
+    (with-redefs [shared/file-exists? (constantly false)]
+      (f.file-management/did-close "file:///aaa.clj" db/db*))
+    (let [db @db/db*]
+      (is (empty? (get-in db [:dep-graph 'aaa :dependencies])))
+      (is (= '{ccc 1} (get-in db [:dep-graph 'aaa :dependents]))) ;; <-- no change, because ccc stil depends on aaa, even though aaa is now undefined
```

Comment:
good corner case, I wonder how that would affect us, like how bad is to store a ns dependent/dependency that its definition doesn't exist anymore

Reply:
I know this is not the case this `is` is checking, but it made me thing about that

Reply:
I've thought a little about what would happen if the dep graph gets corrupted. Obviously that _shouldn't_ happen, but if it does, what would be the result?

There are probably too many cases to consider fully. I don't think anything bad would happen if we store extra data, like relationships that don't actually exist anymore, or files that don't actually define a namespace. The reason is that whenever we have a namespace and a file that defines it, we can look up that file's analysis, but there's always the possibility that the file defines two or more namespaces. So, when we iterate through that file's elements we always (I think) double check that their `:from` is the namespace we're interested in. If we were storing extra data, I think this second check of namespace would eliminate extraneous elements.

If we store too _little_ data, i.e. miss a relationship or a file, the queries would return bad results. Queries wouldn't pull in the right files' analysis, and we'd return too few elements.

Of course I don't _think_ that the dep-graph can get out of sync with the analysis, and so I think we're safe. But, if it does, that's one of the things I'm hoping would turn up in beta testing.